clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java

```diff
@@ -880,27 +884,38 @@ public void handle(LeaveGroupResponse leaveResponse, RequestFuture<Void> future)
         }
     }
 
+    public void setHeartbeatCallbacks(ConsumerHeartbeatDataCallbacks callbacks) {
+        this.consumerHeartbeatDataCallbacks = callbacks;
+    }
+
     // visible for testing
-    synchronized RequestFuture<Void> sendHeartbeatRequest() {
+    synchronized RequestFuture<Map<String,ByteBuffer>> sendHeartbeatRequest() {
         log.debug("Sending Heartbeat request to coordinator {}", coordinator);
         HeartbeatRequest.Builder requestBuilder =
                 new HeartbeatRequest.Builder(new HeartbeatRequestData()
                         .setGroupId(rebalanceConfig.groupId)
                         .setMemberId(this.generation.memberId)
                         .setGroupInstanceId(this.rebalanceConfig.groupInstanceId.orElse(null))
-                        .setGenerationId(this.generation.generationId));
+                        .setGenerationId(this.generation.generationId)
+                        .setUserData(getHeartbeatData())
```

Comment:
Seems some changes in `HeartbeatRequest` was not included?

Reply:
yeah, I didn't realize I'd edited a generated class.